#### Topic:      Prework Solutions
#### Cohort:   03_Spring2015
#### Date:        01-Apr-2015
#### Author:     Reshama Shaikh

Prework: Required Exercises

https://github.com/datascopeanalytics/metis-data-science-bootcamp-prework/blob/master/exercises.md

Source:     Think Stats (section Using the Code), there is some accompanying code and data. You can get these from the Think Stats repository.
https://github.com/AllenDowney/ThinkStats2

------------------------------------------------------------------------------- 

####Note:

Step1.  Create a directory on your computer.  Below is an example:

/Users/reshamashaikh/ds/metis/metisgh/

Step2.  use GitHub to pull this repo to your computer

git clone https://github.com/AllenDowney/ThinkStats2.git

Step3.  Put your ipython notebook in this directory (that way, it can pull the needed dependencies):  

/Users/reshamashaikh/ds/metis/metisgh/ThinkStats2/code/

(content will match:  https://github.com/AllenDowney/ThinkStats2/tree/master/code )

Step4.  Call your prework solutions notebook:  

0_S_Prework_myname.ipynb

##Table of Contents 
[Exercise01](#1) 

(Think Stats Ex 2.4) 

Using the variable totalwgt_lb, investigate whether first babies are lighter or heavier than others. Compute Cohen’s d to quantify the difference between the groups. How does it compare to the difference in pregnancy length?
 

[Exercise02](#2)

(Think Stats Ex 3.1) 

Something like the class size paradox appears if you survey children and ask how many children are in their family. Families with many children are more likely to appear in your sample, and families with no children have no chance to be in the sample.

Use the NSFG respondent variable NUMKDHH to construct the actual distribution for the number of children under 18 in the household.
Now compute the biased distribution we would see if we surveyed the children and asked them how many children under 18 (including themselves) are in their household.

Plot the actual and biased distributions, and compute their means. As a starting place, you can use chap03ex.ipynb. This is an ipython notebook from the ThinkStats2 repository.


[Exercise03](#3)

(Think Stats Ex 4.2) 

The numbers generated by random.random are supposed to be uniform between 0 and 1; that is, every value in the range should have the same probability.
Generate 1000 numbers from random.random and plot their PMF and CDF. Is the distribution uniform?



[Exercise04](#4)

(Think Stats Ex 7.1) 

Using data from the NSFG, make a scatter plot of birth weight versus mother’s age. Plot percentiles of birth weight versus mother’s age. Compute Pearson’s and Spearman’s correlations. How would you characterize the relationship between these variables?

[Exercise05](#5)

(Think Stats Ex 8.2)

Suppose that you draw a sample with size n = 10 from an exponential distribution with λ = 2. Simulate this experiment 1000 times and plot the sampling distribution of the estimate L. Compute the standard error of the estimate and the 90% confidence interval.

Repeat the experiment with a few different values of n and make a plot of standard error versus n.


[Exercise06](#6)

(Think Stats Ex 2.1 Bayes)

The cookie problem is a problem discussed in sections 1.3, 2.2 and 2.3 of Think Bayes. Solve the following problem. In Section 2.3 I said that the solution to the cookie problem generalizes to the case where we draw multiple cookies with replacement. But in the more likely scenario where we eat the cookies we draw, the likelihood of each draw depends on the previous draws.

Modify the solution in this chapter to handle selection without replacement. Hint: add instance variables to Cookie to represent the hypothetical state of the bowls, and modify Likelihood accordingly. You might want to define a Bowl object.


-----------------------------------------------------------------------------------------

### <a id='1'></a> Exercise 1 
(Think Stats Ex 2.4)

a) Using the variable totalwgt_lb, investigate whether first babies are lighter or heavier than others. 

b) Compute Cohen’s d to quantify the difference between the groups. 

c) How does it compare to the difference in pregnancy length?

Note:  Cohen's d is an effect size used to indicate the standardised difference between two means. It can be used, for example, to accompany reporting of t-test and ANOVA results. 

In [81]:
# set working directory
path_data = "/Users/reshamashaikh/_ds/metis/metisgh/ThinkStats2"

In [82]:
import nsfg
df = nsfg.ReadFemPreg()
print "Length of df: ", len(df)
df.head()

Length of df:  13593


caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0       1         1        NaN        NaN       NaN       NaN         6   
1       1         2        NaN        NaN       NaN       NaN         6   
2       2         1        NaN        NaN       NaN       NaN         5   
3       2         2        NaN        NaN       NaN       NaN         6   
4       2         3        NaN        NaN       NaN       NaN         6   

   pregend2  nbrnaliv  multbrth    ...     laborfor_i  religion_i  metro_i  \
0       NaN         1       NaN    ...              0           0        0   
1       NaN         1       NaN    ...              0           0        0   
2       NaN         3         5    ...              0           0        0   
3       NaN         1       NaN    ...              0           0        0   
4       NaN         1       NaN    ...              0           0        0   

       basewgt  adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw  \
0  3410.389399      3869.349602   6448.271112       2     9      NaN   
1  3410.389399      3869.349602   6448.271112       2     9      NaN   
2  7226.301740      8567.549110  12999.542264       2    12      NaN   
3  7226.301740      8567.549110  12999.542264       2    12      NaN   
4  7226.301740      8567.549110  12999.542264       2    12      NaN   

   totalwgt_lb  
0       8.8125  
1       7.8750  
2       9.1250  
3       7.0000  
4       6.1875  

[5 rows x 244 columns]

In [83]:
firsts = df[df.birthord==1]
others = df[df.birthord>1]
print "len of firsts: ", len(firsts)
print "len of others: ", len(others)

len of firsts:  4413
len of others:  4735


In [84]:
print "firsts total birth wt (mean): ", firsts.totalwgt_lb.mean()
print "others total birth wt (mean): ", others.totalwgt_lb.mean()

firsts total birth wt (mean):  7.20109443044
others total birth wt (mean):  7.32585561497


####Question: 1a) Using the variable totalwgt_lb, investigate whether first babies are lighter or heavier than others.
####Answer:   
####First babies are lighter than others (7.20 vs 7.33 lbs)

In [85]:
#b) Compute Cohen’s D to quantify the difference between the groups. 

# Cohen's D function

def CohenD(grp1, grp2):
    print "Computing Cohen D statistic"
    
    
    
    
    
    
    
    

####Answer:  

####Cohen's D is:  -0.0887

c) How does Cohen's D compare to the difference in pregnancy length? 

In [86]:
print "Cohen's D for birth weight is:      x.xxxx"
print "Cohen's D for pregnancy length is:  x.xxxx"

Cohen's D for birth weight is:      x.xxxx
Cohen's D for pregnancy length is:  x.xxxx


####Interpretation
This site has a good interpretation of Cohen's D, effect size:
http://www.uccs.edu/lbecker/effect-size.html

[fill in ] 

-------------------------------------------------------------------------------------------

### <a id='2'></a> Exercise 2 
(Think Stats Ex 3.1)

Something like the class size paradox appears if you survey children and ask how many children are in their family. Families with many children are more likely to appear in your sample, and families with no children have no chance to be in the sample.

Use the NSFG respondent variable NUMKDHH to construct the actual distribution for the number of children under 18 in the household.

Now compute the biased distribution we would see if we surveyed the children and asked them how many children under 18 (including themselves) are in their household.

Plot the actual and biased distributions, and compute their means. As a starting place, you can use chap03ex.ipynb. This is an ipython notebook from the ThinkStats2 repository.

In [1]:
import chap01soln
import thinkstats2
import thinkplot

resp = chap01soln.ReadFemResp()
pmf = thinkstats2.Pmf(resp.numkdhh)
#thinkplot.Pmf(pmf, label = 'numkdhh')
#thinkplot.Show(xlabel = 'Number of Children', ylabel = 'Frequency')

def BiasPmf(pmf, label = ''):
	new_pmf = pmf.Copy(label = label)
	for x, p in pmf.Items():
		new_pmf.Mult(x,x)
	new_pmf.Normalize()
	return new_pmf

new_pmf = BiasPmf(pmf, label = "Biased")


thinkplot.PrePlot(2)
thinkplot.Pmfs([pmf, new_pmf])
thinkplot.Show(xlabel="Number of Children", ylabel="PMF")

print 'Mean of families with children ' ,pmf.Mean()
print 'Mean of all families ' ,new_pmf.Mean()

Mean of families with children  1.02420515504
Mean of all families  2.40367910066


-------------------------------------------------------------------------------------------

### <a id='3'></a> Exercise 3 
(Think Stats Ex 4.2)

The numbers generated by random.random are supposed to be uniform between 0 and 1; that is, every value in the range should have the same probability.
Generate 1000 numbers from random.random and plot their PMF and CDF. Is the distribution uniform?

In [19]:
import thinkstats2
 
import first
import thinkplot
import random


#live, firsts, others = first.MakeFrames()
#cdf = thinkstats2.Cdf(live.totalwgt_lb, label="totalwgt_lb")
#otherCdf = thinkstats2.Cdf(others.totalwgt_lb, label="totalwgt_lb")
#thinkplot.Cdf(cdf)
#thinkplot.Show(xlabel="weeks", ylabel="CDF")

#me = otherCdf.PercentileRank(8.0)
#print "My rank: ", me
nums = []
for i in range (1000):
	nums.append(random.random())

numsCdf = thinkstats2.Cdf(nums, label = "Random Numbers CDF")
thinkplot.Cdf(numsCdf)
thinkplot.Show()

numsPmf = thinkstats2.Pmf(nums, label = "Random Numbers PMF")
thinkplot.Pmf(numsPmf)
thinkplot.Show()

####Answer:
Yes, distribution looks uniform

-------------------------------------------------------------------------------------------

### <a id='4'></a> Exercise 4 
(Think Stats Ex 7.1)

Using data from the NSFG, make a scatter plot of birth weight versus mother’s age. Plot percentiles of birth weight versus mother’s age. Compute Pearson’s and Spearman’s correlations. How would you characterize the relationship between these variables?

In [33]:
import nsfg
import thinkstats2
import thinkplot
import numpy as np 
import first
import sys
import math
import pandas	


def Scatter(ages, weights):
	thinkplot.Scatter(ages, weights)
	thinkplot.Show( xlabel="Mother's Age", 
			   		ylabel="Birth Weight",
			   		axis=[10,50,0,15])

def BinnedPercentiles(df):
	
	bins = np.arange(10,48,3)
	indices = np.digitize(df.agepreg, bins)
	groups = df.groupby(indices)

	ages = [group.agepreg.mean() for i, group in groups][1:-1]
	#print ages
	cdfs = [thinkstats2.Cdf(group.totalwgt_lb) for i, group in groups][1:-1]
	thinkplot.PrePlot(3)
	for percent in [75, 50, 25]:
		weights = [cdf.Percentile(percent) for cdf in cdfs]
		label = '%dth' % percent
		thinkplot.Plot(ages,weights, label=label)
	thinkplot.show(xlabel="Mother's age",
		ylabel ="Baby's Weight",
		axis=[15,45,6.0,8.5])

def covar(x,y,meanx=None,meany=None):
	xs = np.asarray(x)
	ys = np.asarray(y)
	if meanx is None:
		meanx = np.mean(xs)
	if meany is None:
		meany = np.mean(ys)
	cov = np.dot(xs-meanx, ys-meany) / len(xs)
	return cov

def PearsonsCorr(x,y):
	x = np.asarray(x)
	y = np.asarray(y)
	meanx = np.mean(x)
	meany = np.mean(y)
	varx = np.var(x)
	vary = np.var(y)

	corr = covar(x,y, meanx,meany) / math.sqrt(varx * vary)
	return corr

def SpearmanCorr(xs, ys):
    xranks = pandas.Series(xs)
    yranks = pandas.Series(ys)
    return xs.corr(ys, method='spearman')

thinkstats2.RandomSeed(17)
live, firsts, others = first.MakeFrames()
live = live.dropna(subset=['agepreg', 'totalwgt_lb'])
BinnedPercentiles(live)

ages = live.agepreg
weights = live.totalwgt_lb

pearson = PearsonsCorr(ages, weights)
spearman = SpearmanCorr(ages, weights)
print ("Pearson's Correlation: ", pearson)
print ("Spearman's Correlation: ", spearman)

Scatter(ages, weights)





Pearson's Correlation:  0.0688339703541
Spearman's Correlation:  0.0946100410966


####Answer:
For the Pearson r correlation, both variables should be normally distributed.  Other assumptions include linearity and homoscedasticity.  Linearity assumes a straight line relationship between each of the variables in the analysis and homoscedasticity assumes that data is normally distributed about the regression line.

Spearman rank correlation test does not make any assumptions about the distribution. 

####Conclusions:
Younger mothers <20 have lightest babies
Weights seem to stabilize /plateau around 25-35
Weights decrease a bit randomly at >35

Technical definition of low weight baby <5.5

Both Spearman & Pearson correlations are positive and small. The relationship is probably non-linear. 


-----------------------------------------------------------------------------------------

### <a id='5'></a> Exercise 5
(Think Stats Ex 8.2)

Suppose that you draw a sample with size n = 10 from an exponential distribution with λ = 2. Simulate this experiment 1000 times and plot the sampling distribution of the estimate L. Compute the standard error of the estimate and the 90% confidence interval.

Repeat the experiment with a few different values of n and make a plot of standard error versus n.

In [22]:
from __future__ import print_function

import thinkstats2
import thinkplot

import math
import random
import numpy as np
import estimation

def Estimate3(n=10, m=10):


	def VertLine(x, y=1):
		thinkplot.Plot([x, x], [0, y], color='0.8', linewidth=3)

	lam = 2
	means = []
	medians = []
	for _ in range(m):
		xs = np.random.exponential(1.0/lam, n)
		L = 1 / np.mean(xs)
		Lm = math.log(2) / thinkstats2.Median(xs)
		means.append(L)
		medians.append(Lm)


	cdf = thinkstats2.MakeCdfFromList(means)
	ci = cdf.Percentile(5), cdf.Percentile(95)
	print("Confidence Interval",ci)
	VertLine(ci[0])
	VertLine(ci[1])
	thinkplot.Cdf(cdf)
	thinkplot.Show()
	

	print('rmse L', estimation.RMSE(means, lam))
	print('rmse Lm', estimation.RMSE(medians, lam))
	print('mean error L', estimation.MeanError(means, lam))
	print('mean error Lm', estimation.MeanError(medians, lam))


Estimate3(10, 500)

Confidence Interval (1.2559445099031703, 3.4818611972243843)
rmse L 0.803524840175
rmse Lm 1.7580807999
mean error L 0.194322294378
mean error Lm 0.690432044536


-----------------------------------------------------------------------------------------

### <a id='6'></a> Exercise 6
(Think Stats Ex 2.1 Bayes)

The cookie problem is a problem discussed in sections 1.3, 2.2 and 2.3 of Think Bayes. Solve the following problem. In Section 2.3 I said that the solution to the cookie problem generalizes to the case where we draw multiple cookies with replacement. But in the more likely scenario where we eat the cookies we draw, the likelihood of each draw depends on the previous draws.

Modify the solution in this chapter to handle selection without replacement. Hint: add instance variables to Cookie to represent the hypothetical state of the bowls, and modify Likelihood accordingly. You might want to define a Bowl object.

In [90]:
# get this file from:  https://github.com/AllenDowney/ThinkBayes/blob/master/thinkbayes.py
from thinkbayes import Pmf

In [ ]:
Exercise 5.1 In the BRFSS, the distribution of heights is
roughly normal with parameters m = 178 cm and s = 7.7 cm for men,
and m = 163 cm and s = 7.3 cm for women.
In order to join Blue Man Group, you have to be male between 5’10” and
6’1” (see http://bluemancasting.com). What percentage of the
        
min=dist.cdf(177.8)
max=dist.cdf(185.42)
print max-min
0.342746837631
Approximately 34% of the population could apply.
